### DISCUSSION TITLE
[Question] is it possible to force wpp to the same name mangling as wcc?

### DISCUSSION THREAD

**Discussion by LowLevelMahn** [open-watcom/open-watcom-v2#843] [Q&A] [answered]
`void __cdecl __near cc_GAME_START_sub_3(uint16_t es_, uint16_t di_, uint16_t ds_, uint16_t si_)`

wcc name mangling:

`_cc_GAME_START_sub_3`

wpp name mangling:

`W?cc_GAME_START_sub_3$N_USUSUSUS_V`

is it possible to force wpp to also use the wcc style? (for all or better only this function thats gets used in assembler code)

(future: __cdecl and __near gets removed when my reversing project is 100% ported C/C++ code and then the mangling is also not relevant anymore)

**Comment by jmalak** [comment]
Are you sure that you understand what you are asking for?
Any C++ compiler has its own mangling schema to create unique symbol for each class member.
Such C++ mangling must be orthogonal to C compiler mangling because C++ compiler can use both **`C++`** or **`C`** linkage
See [C++ mangling](https://github.com/open-watcom/open-watcom-v2/blob/master/bld/plusplus/y/mangle.y)

**Comment by LowLevelMahn** [comment]
> Are you sure that you understand what you are asking for?

in a every-C++-project context: this can produce evil name clashes due to overloading feature of C++ etc.

in my local context yes - and im talking about decent free functions - for the asm/C++ interaction,  garanteed differently named - not the complete program

i ask different: 

i would like to use the wpp compiler for my reversing (because i hate to predefine my local variables in C and reversed code contains masses of them)

over the time: partial routines in assembler, partially in C++ - temporary - until everything is ported and fully tested - in the end only C++ will survive - currently im doing that usually in C - but i want to try C++ for better readability

if i use wcc i just write

`extern C cc_GAME_START_sub_3:near`

in my assembler source

but that failed with wpp (due to the C++ mangling)

i've tried

`extern W?cc_GAME_START_sub_3$N_USUSUSUS_V:near`

but that gives an error with wlink `Error! E2028: __near cc_GAME_START_sub_3 is an undefined reference` 
and its looks ugly - my next project contains 620 functions - that will look messy

but maybe its a just too special what im asking for

**Comment by jmalak** [comment, marked as answer]
Generally on Assembly level you don't know what is C and what is C++ symbol.
C++ symbols mangling is depend on used compiler against C where some rules are defined for cdecl.
But OW doesn't use cdecl as default therefore C names use mangling rules depend on calling convention which can be different symbol-by-symbol.
If you are sure C++ compiler was OW or Watcom then you can use demangling functions which is included in C run-time-library. 
But you need to know which symbol is C++ and which is C.

I am not sure but `assembler has different character set` for symbol names then C or C++ language that you can have a problem with some symbol names. for such names you can use back quote against such symbol name in wasm assembler.

What linker show you is unmangled C++ symbol, if you want to see native symbol then use linker option
`option MANGlednames`
But it has nothing to do with unresolved symbols.
If you get this message then you must check what symbol name is used by assembly code.

**Comment by LowLevelMahn** [comment]
> What linker show you is unmangled C++ symbol, if you want to see native symbol then use linker option `option MANGlednames` But it has nothing to do with unresolved symbols. If you get this message then you must check what symbol name is used by assembly code.

with `MANGlednames` the error is: 

`Error! E2028: W?cc_GAME_START_sub_3$N_USUSUSUS_V is an undefined reference`

the name in the obj file is (with dmpobj.exe)

```
PUBDEF(90) recnum:19, offset:00000174h, len:003dh, chksum:1bh(1b)
    Group: 0, Seg: 1
    00000000h - 'W?get_ivt$N()PFI' Type:0
    00000005h - 'W?cc_GAME_START_sub_3$N(USUSUSUS)V' Type:0
```

IDA Pro resolves it to (IDA directly supports omf/coff obj files)

`W?cc_GAME_START_sub_3$N_USUSUSUS_V`

so it could be a problem with `assembler has different character set` for symbol names then C or C++ language

**Reply by jmalak** [reply]
If you have coff or omf file then you can use wdis to get proper symbol names.

**Reply by LowLevelMahn** [reply]
wdis -m says sames as dmpobj

`W?cc_GAME_START_sub_3$N(USUSUSUS)V`

i think wasm does not like the parenthesis

`extern W?cc_GAME_START_sub_3$N(USUSUSUS)V:proc`

wasm gives 'Error! E214: Colon is expected'

any docs about how to let wasm accept the parenthesis?

**Reply by jmalak** [reply]
As I wrote you must use back quote for such names, especialy any type of parentheses are not regular character for symbol name.

**Reply by LowLevelMahn** [reply]
thanks - that works

**Reply by jmalak** [reply]
Anyway wdis should use back quote automaticaly for such symbols, there is special function which check all characters in symbol name and if some funny character is used then back quote is inserted automaticaly. If it doesn't work correctly then report it as issue.

**Reply by LowLevelMahn** [reply]
test.cpp
`void test(short a, int b, long c){}
`
wpp.exe test.cpp

wdis.exe -m test.obj

`W?test$N(SIL)V`

no quoting

wdis.exe test.obj

`void __near test( short, int, long )`

i tried to use back quoting in this asm example but it does not work
using back quote, back slash etc. (i thought i did checked it)

asm.asm

```
;extern W?test$N(SIL)V:near
;extern W`?test`$N`(SIL`)V:near
;extern W\?test\$N\(SIL\)V:near

END
```
wasm.exe asm.asm

INFO: i solved the initial problem by implementing plain extern "C" functions in cpp - that is ok for me
just some more parameter, but at least no upfront-definition of local variables :)

**Reply by jmalak** [reply]
I see, the problem is that you use incorrect options for wdis.
Proper is `wdis -m -a test.obj`

**Reply by LowLevelMahn** [reply]
-a quotes the complete name and that works in wasm - thanks

**Comment by pchapin** [comment]
Isn't this what extern "C" is for? In your C++ code do:

    extern "C" {
        void some_function( int x );
    }

The C++ compiler will not use C++ name mangling on such functions so they end up with names that are compatible with the names produced by the C compiler. Typically this is used for calling C functions from C++, but you can do this even for functions written in C++ to make them callable by C. However, calling C++ from C might be trickier because of the runtime library requirements.

**Reply by LowLevelMahn** [reply]
you're right: my first idea was to use C++ classes/methods, but in the end i fully switched to extern C functions - but also due to problems with wpp and code-segment variables, the game code im partially porting - oder better said try to half port to C/C++ when the original code is abusing ds, cs, sometimes even ss register etc. is very hard
